walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -20,44 +20,87 @@
-      for (TimeRange meetingTime : meetingTimes) {
-        if (eventTime.overlaps(meetingTime)) {
-          meetingTimes.remove(meetingTime);
-          if (meetingTime.start() < eventTime.start()) {
-            TimeRange before = TimeRange.fromStartEnd(meetingTime.start(), eventTime.start(), false);
-            if (before.duration() >= request.getDuration()) {
-              meetingTimes.add(before);
-            }
-          }
-          if (eventTime.end() < meetingTime.end()) {
-            TimeRange after = TimeRange.fromStartEnd(eventTime.end(), meetingTime.end(), false);
-            if (after.duration() >= request.getDuration()) {
-              meetingTimes.add(after);
-            }
-          }
-        }
+      meetingTimes = splitMeetingTimes(event, meetingTimes, meetingDuration);
+      if (meetingTimes.isEmpty()) {
+        break;
+      }
+    }
+
+    Collection<TimeRange> optionalMeetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
+    optionalMeetingTimes.addAll(meetingTimes);
+    for (Event optionalEvent : optionalEvents) {
+      optionalMeetingTimes = splitMeetingTimes(optionalEvent, optionalMeetingTimes, meetingDuration);
+      if (optionalMeetingTimes.isEmpty()) {
+        break;
       }
     }
+
+    if (mandatoryMeetingAttendees.isEmpty()) {
+      meetingTimes = optionalMeetingTimes;
+    } else {
+      meetingTimes = optionalMeetingTimes.isEmpty() ? meetingTimes : optionalMeetingTimes;
+    }
+
     Collection<TimeRange> finalMeetingTimes = new ArrayList<>();
     finalMeetingTimes.addAll(meetingTimes);
     return finalMeetingTimes;
   }
+
+  public Collection<TimeRange> splitMeetingTimes(Event event, Collection<TimeRange> meetingTimes, long meetingDuration) {
+    Collection<TimeRange> newMeetingTimes = new TreeSet<>(TimeRange.ORDER_BY_START);
+    TimeRange eventTime = event.getWhen();
+    for (TimeRange meetingTime : meetingTimes) {
+      newMeetingTimes = updateMeetingTimes(newMeetingTimes, eventTime, meetingTime, meetingDuration);
+    }
+    return newMeetingTimes;
+  }
+
+  public Collection<TimeRange> updateMeetingTimes(Collection<TimeRange> meetingTimes, TimeRange eventTime, TimeRange meetingTime, long meetingDuration) {
+    if (eventTime.overlaps(meetingTime)) {
+      meetingTimes = addMeetingTime(meetingTimes, meetingTime.start(), eventTime.start(), meetingDuration);
+      meetingTimes = addMeetingTime(meetingTimes, eventTime.end(), meetingTime.end(), meetingDuration);
+    } else {
+      meetingTimes.add(meetingTime);
+    }
+    return meetingTimes;
+  }
+
+  public Collection<TimeRange> addMeetingTime(Collection<TimeRange> meetingTimes, int startTime, int endTime, long meetingDuration) {
+    Optional optionalTime = newMeetingTime(startTime, endTime, meetingDuration);
+    optionalTime.ifPresent(timeRange -> meetingTimes.add((TimeRange) timeRange));
+    return meetingTimes;
+  }
+
+  public Optional<TimeRange> newMeetingTime(int startTime, int endTime, long meetingDuration) {
```

Comment:
[Optional] This function I believe is only used once so it would be fine to remove it and just put this logic directly in the addMeetingTime function.  The question of when to abstract a method can be challenging but I tend to do it if one of 3 things are true
1) Identical or similar code is used more than once
2) A single method is too many lines of code ('too many' usually meaning for me that I can't see it all at once on my monitor)
3) There is a logic chunk that can be clearly abstracted.

In this case #1 and #2 are clearly false, but I think #3 is debatable so I'll leave it up to you whether you want to keep the new method or remove it.

Reply:
Wow, it linked to old PRs when I added the '#' sign lol.  Ignore those auto links.

Reply:
I agree with you, I believe that function isn't necessary.